vpr/src/place/analytic_placer.cpp

```diff
@@ -0,0 +1,705 @@
+            ClusterPinId pin_id = clb_nlist.net_pin(net_id, ipin);
+            auto process_arc = [&](ClusterPinId other) {
+                if (other == pin_id || pin_id == ubpin) // to avoid bound pins connect to itself && double counting lbpin-to-ubpin connection
+                    return;
+                int o_pos = blk_p(clb_nlist.pin_block(other));
+                double weight = 1.0 / ((pin_n - 1) * std::max<double>(1, std::abs(o_pos - this_pos)));
+
+                // TODO: adding timing weights to matrix entries
+                /*if (ipin != 0){
+                 * weight *= (1.0 + tmpCfg.timingWeight * std::pow(place_crit.criticality(net_id, ipin), tmgCfg.criticalityExponent));
+                 * } */
+
+                stamp_equation(pin_id, pin_id, weight);
+                stamp_equation(pin_id, other, -weight);
+                stamp_equation(other, other, weight);
+                stamp_equation(other, pin_id, -weight);
+            };
+            process_arc(lbpin);
+            process_arc(ubpin);
+        }
+    }
+    if (iter != -1) {
+        for (size_t row = 0; row < solve_blks.size(); row++) {
+            int l_pos = legal_p(solve_blks.at(row));
+            int b_pos = blk_p(solve_blks.at(row));
+
+            double weight = tmpCfg.alpha * iter / std::max<double>(1, std::abs(l_pos - b_pos));
+            // Add psudo-connections for legalization
+            es.add_coeff(row, row, weight);
+            es.add_rhs(row, weight * l_pos);
+        }
+    }
+}
+
+// Solve the system of equations
+void AnalyticPlacer::solve_equations(EquationSystem<double>& es, bool yaxis) {
+    int max_x = g_vpr_ctx.device().grid.width();
+    int max_y = g_vpr_ctx.device().grid.height();
+
+    auto blk_pos = [&](ClusterBlockId blk_id) { return yaxis ? blk_locs.at(blk_id).loc.y : blk_locs.at(blk_id).loc.x; };
+    std::vector<double> vals;
+    std::transform(solve_blks.begin(), solve_blks.end(), std::back_inserter(vals), blk_pos);
+    es.solve(vals, tmpCfg.solverTolerance);
+
+    for (size_t i = 0; i < vals.size(); i++)
+        if (yaxis) {
+            blk_locs.at(solve_blks.at(i)).rawy = vals.at(i);
+            blk_locs.at(solve_blks.at(i)).loc.y = std::min(max_y - 1, std::max(0, int(vals.at(i) + 0.5)));
+        } else {
+            blk_locs.at(solve_blks.at(i)).rawx = vals.at(i);
+            blk_locs.at(solve_blks.at(i)).loc.x = std::min(max_x - 1, std::max(0, int(vals.at(i) + 0.5)));
+        }
+}
+
+// Debug use, finds # of blocks on 1 tile for all tiles
+float AnalyticPlacer::find_overlap() {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+    int max_x = g_vpr_ctx.device().grid.width();
+    int max_y = g_vpr_ctx.device().grid.height();
+
+    int OL = 0;
+    overlap.resize(max_y);
+    for (auto& row : overlap) {
+        row.resize(max_x);
+        std::fill(row.begin(), row.end(), 0);
+    }
+
+    for (auto blk : clb_nlist.blocks()) {
```

Comment:
the vtr::Matrix class would be a good fit to this variable (cleaner).  See https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/master/libs/libvtrutil/src/vtr_ndmatrix.h if you want to see the code, but probably easiest to look for an example of its use (which is easy; behaves like:
my_matrix[i][j] = 20; // Access as if it's a 2D array.